### PR TITLE
Problem: (CE-267) Missing API documentation

### PR DESCRIPTION
**Pull request by calvinlauco**
Solution: Added TypeDoc documentation generator

-----

Generate API documentation by running
```bash
npm run docs:generate
```

Generated documentation is available at `docs/dist`

### CODE REVIEW

**Comment by CeruleanAtMonaco** [comment]
which node version does this work with? 
I can't install the packages with node v13.11.0

**Comment by calvinaco** [comment]
> which node version does this work with?
> I can't install the packages with node v13.11.0

It should work on at least most of the major version. Our previous Ci tested with Node.js 8, 9 10 and 14. Let me try with v13. May I know what error did you got?

**Comment by cdc-Hitesh** [comment]
@calvinaco  @CeruleanAtMonaco 
Even I happen to run into the module installation problem on v10 or v14. Seems it was a problem with the `gyp` tool on MacOS.

I followed this tutorial (https://medium.com/flawless-app-stories/gyp-no-xcode-or-clt-version-detected-macos-catalina-anansewaa-38b536389e8d) and then installed packages.

- [ ]  `npm i grunt` --> as it was missing somehow

- [ ]  `cd ./docs/themes/ && npm i` --> was throwing local package.json error so had to do this

- [ ]  `cd ../../` and run  `npm run docs:generate`

I am not sure but that's how I was only able to generate a `dist` folder with docs.

**Comment by CeruleanAtMonaco** [comment]
> > which node version does this work with?
> > I can't install the packages with node v13.11.0
> 
> It should work on at least most of the major version. Our Ci covered Node.js 8, 9 10 and 14. Let me try with v13. May I know what error did you got?

It's okay, I can switch to 14

**Comment by CeruleanAtMonaco** [comment]
> > > which node version does this work with?
> > > I can't install the packages with node v13.11.0
> > 
> > 
> > It should work on at least most of the major version. Our Ci covered Node.js 8, 9 10 and 14. Let me try with v13. May I know what error did you got?
> 
> It's okay, I can switch to 14

just for context, the error I got:

```
chain-nodelib git:(feature/cd-267-apidoc) npm run docs:generate

> @crypto-com/chain-nodelib@0.3.1 docs:generate /Users/ceruleanhu/code/chain/chain-nodelib
> ./typedoc.sh

Generating documentation for version 0.3.1

> @crypto-com/chain-nodelib@0.3.1 typedoc:generate /Users/ceruleanhu/code/chain/chain-nodelib
> npm run typedoc:theme && typedoc --options typedoc.json "--name" "Node.js Library v0.3.1" "--gaID" ""


> @crypto-com/chain-nodelib@0.3.1 typedoc:theme /Users/ceruleanhu/code/chain/chain-nodelib
> cd docs/themes && npm run build


> typedoc-default-themes@0.10.2 build /Users/ceruleanhu/code/chain/chain-nodelib/docs/themes
> npm run clean && npm run grunt


> typedoc-default-themes@0.10.2 clean /Users/ceruleanhu/code/chain/chain-nodelib/docs/themes
> rimraf bin

sh: rimraf: command not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! typedoc-default-themes@0.10.2 clean: `rimraf bin`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the typedoc-default-themes@0.10.2 clean script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ceruleanhu/.npm/_logs/2020-08-25T04_47_30_272Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! typedoc-default-themes@0.10.2 build: `npm run clean && npm run grunt`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the typedoc-default-themes@0.10.2 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ceruleanhu/.npm/_logs/2020-08-25T04_47_30_294Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @crypto-com/chain-nodelib@0.3.1 typedoc:theme: `cd docs/themes && npm run build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @crypto-com/chain-nodelib@0.3.1 typedoc:theme script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ceruleanhu/.npm/_logs/2020-08-25T04_47_30_314Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @crypto-com/chain-nodelib@0.3.1 typedoc:generate: `npm run typedoc:theme && typedoc --options typedoc.json "--name" "Node.js Library v0.3.1" "--gaID" ""`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @crypto-com/chain-nodelib@0.3.1 typedoc:generate script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ceruleanhu/.npm/_logs/2020-08-25T04_47_30_334Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @crypto-com/chain-nodelib@0.3.1 docs:generate: `./typedoc.sh`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @crypto-com/chain-nodelib@0.3.1 docs:generate script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ceruleanhu/.npm/_logs/2020-08-25T04_47_30_354Z-debug.log
➜  chain-nodelib git:(feature/cd-267-apidoc) npm install

> tiny-secp256k1@1.1.3 install /Users/ceruleanhu/code/chain/chain-nodelib/node_modules/tiny-secp256k1
> npm run build || echo "secp256k1 bindings compilation fail. Pure JS implementation will be used."


> tiny-secp256k1@1.1.3 build /Users/ceruleanhu/code/chain/chain-nodelib/node_modules/tiny-secp256k1
> node-gyp rebuild

  CXX(target) Release/obj.target/secp256k1/native/addon.o
../native/addon.cpp:240:2: warning: ignoring return value of function declared
      with 'warn_unused_result' attribute [-Wunused-result]
        secp256k1_ec_privkey_negate(context, tweak_negated); // returns 1 always
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~
../native/addon.cpp:320:34: error: no viable conversion from
      'v8::Local<v8::Context>' to 'v8::Isolate *'
  ...strict = info[3]->BooleanValue(Nan::GetCurrentContext()).FromJust();
                                    ^~~~~~~~~~~~~~~~~~~~~~~~
/Users/ceruleanhu/Library/Caches/node-gyp/13.11.0/include/node/v8.h:2771:30: note: 
      passing argument to parameter 'isolate' here
  bool BooleanValue(Isolate* isolate) const;
                             ^
../native/addon.cpp:82:36: error: no viable conversion from
      'v8::Local<v8::Context>' to 'v8::Isolate *'
  ...return info[index]->BooleanValue(Nan::GetCurrentContext()).FromJust() ? ...
                                      ^~~~~~~~~~~~~~~~~~~~~~~~
../native/addon.cpp:142:21: note: in instantiation of function template
      specialization '(anonymous namespace)::assumeCompression<2,
      Nan::FunctionCallbackInfo<v8::Value>, v8::Local<v8::Object> >' requested
      here
        const auto flags = assumeCompression<2>(info, pA);
                           ^
/Users/ceruleanhu/Library/Caches/node-gyp/13.11.0/include/node/v8.h:2771:30: note: 
      passing argument to parameter 'isolate' here
  bool BooleanValue(Isolate* isolate) const;
                             ^
../native/addon.cpp:80:21: warning: comparison of integers of different signs:
      'int' and 'unsigned long' [-Wsign-compare]
                if (info.Length() <= index) return __isPointCompressed(p...
                    ~~~~~~~~~~~~~ ^  ~~~~~
../native/addon.cpp:82:36: error: no viable conversion from
      'v8::Local<v8::Context>' to 'v8::Isolate *'
  ...return info[index]->BooleanValue(Nan::GetCurrentContext()).FromJust() ? ...
                                      ^~~~~~~~~~~~~~~~~~~~~~~~
../native/addon.cpp:174:21: note: in instantiation of function template
      specialization '(anonymous namespace)::assumeCompression<1,
      Nan::FunctionCallbackInfo<v8::Value>, v8::Local<v8::Object> >' requested
      here
        const auto flags = assumeCompression<1>(info, p);
                           ^
/Users/ceruleanhu/Library/Caches/node-gyp/13.11.0/include/node/v8.h:2771:30: note: 
      passing argument to parameter 'isolate' here
  bool BooleanValue(Isolate* isolate) const;
                             ^
../native/addon.cpp:80:21: warning: comparison of integers of different signs:
      'int' and 'unsigned long' [-Wsign-compare]
                if (info.Length() <= index) return __isPointCompressed(p...
                    ~~~~~~~~~~~~~ ^  ~~~~~
../native/addon.cpp:89:36: error: no viable conversion from
      'v8::Local<v8::Context>' to 'v8::Isolate *'
  ...return info[index]->BooleanValue(Nan::GetCurrentContext()).FromJust() ? ...
                                      ^~~~~~~~~~~~~~~~~~~~~~~~
../native/addon.cpp:189:21: note: in instantiation of function template
      specialization '(anonymous namespace)::assumeCompression<1,
      Nan::FunctionCallbackInfo<v8::Value> >' requested here
        const auto flags = assumeCompression<1>(info);
                           ^
/Users/ceruleanhu/Library/Caches/node-gyp/13.11.0/include/node/v8.h:2771:30: note: 
      passing argument to parameter 'isolate' here
  bool BooleanValue(Isolate* isolate) const;
                             ^
../native/addon.cpp:87:21: warning: comparison of integers of different signs:
      'int' and 'unsigned long' [-Wsign-compare]
                if (info.Length() <= index) return SECP256K1_EC_COMPRESSED;
                    ~~~~~~~~~~~~~ ^  ~~~~~
4 warnings and 4 errors generated.
make: *** [Release/obj.target/secp256k1/native/addon.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
gyp ERR! stack     at ChildProcess.emit (events.js:315:20)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
gyp ERR! System Darwin 19.6.0
gyp ERR! command "/usr/local/Cellar/node/13.11.0/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/ceruleanhu/code/chain/chain-nodelib/node_modules/tiny-secp256k1
gyp ERR! node -v v13.11.0
gyp ERR! node-gyp -v v5.0.7
gyp ERR! not ok 
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! tiny-secp256k1@1.1.3 build: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the tiny-secp256k1@1.1.3 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ceruleanhu/.npm/_logs/2020-08-25T04_47_50_769Z-debug.log
secp256k1 bindings compilation fail. Pure JS implementation will be used.

> nodemon@2.0.4 postinstall /Users/ceruleanhu/code/chain/chain-nodelib/node_modules/nodemon
> node bin/postinstall || exit 0

Love nodemon? You can now support the project via the open collective:
 > https://opencollective.com/nodemon/donate


> @crypto-com/chain-nodelib@0.3.1 install /Users/ceruleanhu/code/chain/chain-nodelib
> node-pre-gyp install --fallback-to-build=false || npm run build

node-pre-gyp WARN Using needle for node-pre-gyp https download 
node-pre-gyp ERR! install error 
node-pre-gyp ERR! stack Error: 404 status code downloading tarball https://github.com/crypto-com/chain-nodelib/releases/download/0.3.1/node-v79-darwin-x64.tar.gz
node-pre-gyp ERR! stack     at PassThrough.<anonymous> (/Users/ceruleanhu/code/chain/chain-nodelib/node_modules/@amilajack/node-pre-gyp/lib/install.js:149:27)
node-pre-gyp ERR! stack     at PassThrough.emit (events.js:327:22)
node-pre-gyp ERR! stack     at ClientRequest.<anonymous> (/Users/ceruleanhu/code/chain/chain-nodelib/node_modules/needle/lib/needle.js:500:9)
node-pre-gyp ERR! stack     at Object.onceWrapper (events.js:422:26)
node-pre-gyp ERR! stack     at ClientRequest.emit (events.js:315:20)
node-pre-gyp ERR! stack     at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:610:27)
node-pre-gyp ERR! stack     at HTTPParser.parserOnHeadersComplete (_http_common.js:116:17)
node-pre-gyp ERR! stack     at TLSSocket.socketOnData (_http_client.js:479:22)
node-pre-gyp ERR! stack     at TLSSocket.emit (events.js:315:20)
node-pre-gyp ERR! stack     at addChunk (_stream_readable.js:297:12)
node-pre-gyp ERR! System Darwin 19.6.0
node-pre-gyp ERR! command "/usr/local/Cellar/node/13.11.0/bin/node" "/Users/ceruleanhu/code/chain/chain-nodelib/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build=false"
node-pre-gyp ERR! cwd /Users/ceruleanhu/code/chain/chain-nodelib
node-pre-gyp ERR! node -v v13.11.0
node-pre-gyp ERR! node-pre-gyp -v v0.12.0
node-pre-gyp ERR! not ok 
404 status code downloading tarball https://github.com/crypto-com/chain-nodelib/releases/download/0.3.1/node-v79-darwin-x64.tar.gz

> @crypto-com/chain-nodelib@0.3.1 build /Users/ceruleanhu/code/chain/chain-nodelib
> npm run build:neon && npm run build:tsc


> @crypto-com/chain-nodelib@0.3.1 build:neon /Users/ceruleanhu/code/chain/chain-nodelib
> neon build --release

neon info forcing rebuild for new build settings
neon info running cargo
error: failed to parse lock file at: /Users/ceruleanhu/code/chain/chain-nodelib/native/Cargo.lock

Caused by:
  invalid serialized PackageId for key `package.dependencies`
neon ERR! cargo build failed

Error: cargo build failed
    at Target.<anonymous> (/Users/ceruleanhu/code/chain/chain-nodelib/node_modules/neon-cli/lib/target.js:121:35)
    at step (/Users/ceruleanhu/code/chain/chain-nodelib/node_modules/neon-cli/lib/target.js:32:23)
    at Object.next (/Users/ceruleanhu/code/chain/chain-nodelib/node_modules/neon-cli/lib/target.js:13:53)
    at fulfilled (/Users/ceruleanhu/code/chain/chain-nodelib/node_modules/neon-cli/lib/target.js:4:58)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @crypto-com/chain-nodelib@0.3.1 build:neon: `neon build --release`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @crypto-com/chain-nodelib@0.3.1 build:neon script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ceruleanhu/.npm/_logs/2020-08-25T04_47_53_234Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @crypto-com/chain-nodelib@0.3.1 build: `npm run build:neon && npm run build:tsc`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @crypto-com/chain-nodelib@0.3.1 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ceruleanhu/.npm/_logs/2020-08-25T04_47_53_258Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @crypto-com/chain-nodelib@0.3.1 install: `node-pre-gyp install --fallback-to-build=false || npm run build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @crypto-com/chain-nodelib@0.3.1 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ceruleanhu/.npm/_logs/2020-08-25T04_47_53_337Z-debug.log
```

**Comment by calvinlauco** [comment]
Thanks @cdc-Hitesh @CeruleanAtMonaco for the error details. I am able to reproduce the error and has applied a fix. Please delete the branch and pull & checkout again (because I have do a force push)

**Comment by foreseaz** [comment]
Inside `docs/theme` folder, `npm install`  issue seems to still exist. According to this [comment](https://github.com/verzac/typescript-aws-sam/issues/8#issuecomment-666846755), after removing the `package_json.lock` file, I can run install normally. It might due to creating package-json.lock with a different version of NodeJS.
Btw, how can I kick off a dev server locally?

**Comment by calvinlauco** [comment]
> Inside `docs/theme` folder, `npm install` issue seems to still exist. According to this [comment](https://github.com/verzac/typescript-aws-sam/issues/8#issuecomment-666846755), after removing the `package_json.lock` file, I can run install normally. It might due to creating package-json.lock with a different version of NodeJS.
> Btw, how can I kick off a dev server locally? such as run by [grunt-contrib-connect](https://github.com/gruntjs/grunt-contrib-connect). Verified the static outputs `dist` folder, looks good.

Interesting, I think I should create a CI pipeline for each Node.js version to test for the compatibility of the themes folder. May I know which Node.js version you are using?

For the server I was considering a nodemon and simple http server. The grunt plugin also looks good. I could incorporate either solutions on next PR.

File: docs/themes/src/default/assets/css/setup/_grid.sass [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            @content
+
+.root
+    display: flex
+    flex-direction: column
+    min-height: 100vh
+    padding: 0
+
+.container
+    // max-width: 1200px
```
Comment (foreseaz) [inline docs/themes/src/default/assets/css/setup/_grid.sass]: Consider to add `max-width: 1376px;` to be consistent with chain docs?

**Pull request merged**